File: java/base/src/main/java/vsdk/toolkit/numericalAnalysis/quasiMonteCarlo/ScrambledHalton.java
package vsdk.toolkit.numericalAnalysis.quasiMonteCarlo;
/**
Scrambled halton quasiMonteCarlo sample sequence
*/


/**
Scrambled Halton quasiMonteCarlo sequence
*/
public class ScrambledHalton {
    private static final int MAX_DIM = 10;
    private static final int[] prime = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29};
    private static final double[] sample = new double[MAX_DIM];

    /**
    Scrambled halton quasiMonteCarlo sample sequence
    */
    public static double[] scrambledHalton(long nextN, int dim) {
        for (int i = 0; i < dim; i++) {
            int b = prime[i]; // Each component uses different base
            double bi = 1.0 / b;
            double fi = 0;
            double bp = 1;
            int m = 0;

            for (long j = nextN; j > 0; j /= b) {
                bp = bp * bi;
                int a = (int)(j % b); // Variable "a" is m-th digit from b-ary representation of nextN
                a = (a + m) % b; // Permutation of variable "a", Warnock's method
                fi = fi + a * bp;
                m += 1;
            }
            sample[i] = fi;
        }
        return sample;
    }
}

File: typescript/base/src/main/vsdk/toolkit/numericalAnalysis/quasiMonteCarlo/ScrambledHalton.ts
export class ScrambledHalton {
  private static readonly MAX_DIM = 10;
  private static readonly prime = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29];
  private static readonly sample = new Array<number>(ScrambledHalton.MAX_DIM).fill(0.0);

  public static scrambledHalton(nextN: number, dim: number): number[] {
    for (let i = 0; i < dim; i++) {
      const b = ScrambledHalton.prime[i]!;
      const bi = 1.0 / b;
      let fi = 0.0;
      let bp = 1.0;
      let m = 0;

      for (let j = nextN; j > 0; j = globalThis.Math.trunc(j / b)) {
        bp = bp * bi;
        let a = j % b;
        a = (a + m) % b;
        fi = fi + a * bp;
        m += 1;
      }
      ScrambledHalton.sample[i] = fi;
    }

    return ScrambledHalton.sample;
  }

  private constructor() {
  }
}
